Adding update `version` support

NachoPal opened the pull request:
This PR aims to solve the problem discussed in this forum post: https://forum.polkadot.network/t/publish-substrate-to-crates-io/949/39?u=nachopal

It adds two new options to the `update` subcommand:
- `version`: this specifies the source of the versions to which the crates will be updated. There are three valid values:
  - `latest`: it will query the latest crate version from `https://crates.io/`
  - A URL pointing to a raw `Cargo.lock`, for example:
  ```https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/Cargo.lock```
  - A path to a local `Cargo.lock`
- `exclude`: This specifies the path to a TOML file where a `[diener_exclude]` manifest key is expected, listing the crates we do not want to be updated. (This applies to all kinds of updates, not only for `version`)

NachoPal commented:
@bkchr Does it need review/approval or I can hit the merge button?

bkchr commented on src/update.rs:
```diff
@@ -203,10 +312,136 @@ fn handle_toml_file(path: PathBuf, rewrite: &Rewrite, version: &Version) -> Resu
+fn get_version_by_source(package: &str, source: &VersionSource) -> Result<String> {
+    let version = match source {
+        VersionSource::CratesIO => {
+            let url = format!("https://crates.io/api/v1/crates/{}", package);
```
Are you sure that we don't generate too many requests? Maybe we should wait between requests a little bit.

ggwpez replied:
Yes it has limits. I used https://crates.io/data-access which is unlimited [here
](https://github.com/ggwpez/substrate-scripts/blob/4ba5b0df9db832eaecd8d1b507dfc75322c3c795/import-runtime-repos/fix-deps/src/main.rs#L23-L28).  
The create name needs to be split up, but it is not rate-limited as advantage.

NachoPal replied:
> Are you sure that we don't generate too many requests? Maybe we should wait between requests a little bit.

I tested it and it works. Maybe thanks to `reqwest::blocking`?

bkchr reviewed (commented):
Having exclude by passing folders would also be nice.

NachoPal commented on src/update.rs:
```diff
@@ -1,10 +1,21 @@
 use toml_edit::{Document, InlineTable, Value};
 use walkdir::{DirEntry, WalkDir};
 
+thread_local! {
+    /// Packages version - HashMap(name, version)`
+    pub static PACKAGES_VERSION: RefCell<HashMap<String, String>> = RefCell::new(HashMap::new());
+    /// Cargo.lock from URL or FILE sources
+    pub static CARGO_LOCK: RefCell<Option<String>> = RefCell::new(None);
+    /// Excluded packages from version updating - HashMap(name, exclude)`
+    pub static EXCLUDED_PACKAGES: RefCell<HashMap<String, bool>> = RefCell::new(HashMap::new());
+}
```
> Please no thread locals. Remove them please.

Out of curiosity, is there something wrong about using thread locals? is it a bad practice?

NachoPal replied:
Fixed

NachoPal commented:
@bkchr @ggwpez Addressed the comments

In regards
> Having exclude by passing folders would also be nice.

Not sure I understand this. Excluding folders under the same workspace would lead to crates version conflicts.

brenzi commented:
looking forward to this as it simplifies maintenance work for parachain teams.
